crates/namada/src/ledger/protocol/mod.rs

```diff
@@ -621,66 +785,160 @@ fn token_transfer<WLS>(
+        let src_balance = crate::token::read_balance(state, token, src)
+            .map_err(Error::StorageError)?;
+        match src_balance.checked_sub(amount) {
+            Some(new_src_balance) => {
+                let dest_key =
+                    crate::token::storage_key::balance_key(token, dest);
+                let dest_balance =
+                    crate::token::read_balance(state, token, dest)
+                        .map_err(Error::StorageError)?;
+                match dest_balance.checked_add(amount) {
+                    Some(new_dest_balance) => {
+                        state
+                            .write_log_mut()
+                            .write(&src_key, new_src_balance.serialize_to_vec())
+                            .map_err(|e| Error::FeeError(e.to_string()))?;
+                        match state.write_log_mut().write(
+                            &dest_key,
+                            new_dest_balance.serialize_to_vec(),
+                        ) {
+                            Ok(_) => Ok(()),
+                            Err(e) => Err(Error::FeeError(e.to_string())),
+                        }
                     }
+                    None => Err(Error::StorageError(
+                        namada_state::StorageError::new_alloc(format!(
+                            "The transfer would overflow balance of {dest}"
+                        )),
+                    )),
                 }
-                None => Err(Error::FeeError(
-                    "The transfer would overflow destination balance"
-                        .to_string(),
-                )),
+            }
+            None => {
+                Err(Error::StorageError(namada_state::StorageError::new_alloc(
+                    format!("{src} has insufficient balance"),
+                )))
             }
         }
-        None => Err(Error::FeeError("Insufficient source balance".to_string())),
     }
+
+    inner_fee_token_transfer(state, token, src, dest, amount).map_err(|err| {
+        state.write_log_mut().drop_tx();
+
+        err
+    })
 }
 
 /// Check if the fee payer has enough transparent balance to pay fees
-pub fn check_fees<S>(state: &S, wrapper: &WrapperTx) -> Result<()>
+pub fn check_fees<S, D, H, CA>(
+    shell_params: &mut ShellParams<'_, S, D, H, CA>,
+    tx: &Tx,
+    wrapper: &WrapperTx,
+) -> Result<Option<(BatchedTxResult, Hash)>>
 where
-    S: State + StorageRead,
+    S: State<D = D, H = H>
+        + StorageRead
+        + Read<Err = namada_state::Error>
+        + Sync,
+    D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
+    H: 'static + StorageHasher + Sync,
+    CA: 'static + WasmCacheAccess + Sync,
 {
-    let balance = crate::token::read_balance(
-        state,
-        &wrapper.fee.token,
-        &wrapper.fee_payer(),
-    )
-    .unwrap();
-
-    let fees = wrapper
-        .get_tx_fee()
-        .map_err(|e| Error::FeeError(e.to_string()))?;
+    fn inner_check_fees<S, D, H, CA>(
```

Comment:
same question here

Reply:
It's because I want to drop the write log on every error but I wanted to do that in a single place (and also avoid forgetting to call the drop if we need to propagate another error)

Reply:
Actually removed in e1b5857845